src/utils/yup/tests/index.ts

```diff
@@ -0,0 +1,241 @@
+ * call that should pass, the debounced function will return the cached fail value.
+ * @param fn Yup test function.
+ * @param wait Time to wait
+ * @param options Can be passed options from the CustomTestConfig to return early if a certain condition is met,
+ * which can improve performance.
+ * @returns Debounced test function.
+ */
+function yupDebounce(
+  fn: TestFunction,
+  wait: number,
+  options?: YupDebounceOptions,
+) {
+  let lastCallTime;
+  let lastValue;
+  let lastReturnValue;
+  /*
+   * Array to keep track of the number of additions made to the timeout queue.
+   */
+  let timeouts: any[] = [];
+
+  /* Calculate time since last function call. */
+  function shouldInvoke(time) {
+    const timeSinceLastCall = lastCallTime ? time - lastCallTime : 0;
+    return (
+      /*
+       * If lastCallTime is undefined, this is the first time the function has been called.
+       * If timeSinceLastCall is greater than or equal to wait, the delay period has elapsed.
+       * If it's negative, system time has moved backwards. If so, just invoke as if delayed period has elapsed.
+       */
+      lastCallTime === undefined ||
+      timeSinceLastCall >= wait ||
+      timeSinceLastCall < 0
+    );
+  }
+
+  /*
+   * Debounced test function wrapper. If wait is over, invoke function.
+   * Else, queue subsequent calls and only invoke the most recent one.
+   */
+  function caller(value) {
+    /*
+     * If input hasn't changed when called, e.g. if the call is the result of another
+     * form field changing
+     */
+    if (lastReturnValue) {
+      if (Object.is(value, lastValue)) {
+        return lastReturnValue;
+      }
+    }
+    lastValue = value;
+    /*
+     * Return early if circuitbreaker provided and condition satisfied.
+     * Improves performance.
+     */
+    if (options) {
+      const { circuitBreaker, isOptional } = options;
+
+      const cancel = cancelEarly(circuitBreaker, value);
+
+      if (!value || cancel) {
+        lastReturnValue = isOptional;
+        if (!timeouts.length) {
+          return isOptional;
+        }
+        /* Ensures any remaining timeout doesn't resolve after this one */
+        return new Promise((resolve) => {
+          setTimeout(resolve, wait, isOptional);
+        });
+      }
+    }
+    const time = now();
+    const isInvoking = shouldInvoke(time);
+    lastCallTime = time;
+    if (isInvoking) {
+      timeouts = [];
+      const result = fn.call(this, value);
+      lastReturnValue = result;
+      return result;
+    }
+
+    /*
+     * If we're still waiting, return a promise that will resolve after wait has elapsed.
+     * Its resolve value depends on how many timeouts are in the queue.
+     * If there's only one left, resolve with the result of the Yup test function.
+     * Else, resolve as true and remove one timeout from the queue.
+     */
+    const result = new Promise((resolve) => {
+      const id = setTimeout(() => {
+        if (timeouts.length > 1) {
```

Comment:
Could you explain in what case there will be more than one element in `timeouts`?

Reply:
Appreciate the detailed explanation! Can we not just keep track of the most recent call and discard any previous timeouts? This is the pattern I saw the most when doing debouncing, hence my question in the first place. 
Or if this is the best way to implement it, I'm perfectly happy with it. Just curious 🙂

Reply:
If you clear previous timeouts, then the validation will not complete because its promises won't resolve. 

Say you clear the previous timeout on every keystroke in the delay period. You still need to return a promise. But now it never resolves because you cleared the timeout that resolves it. And you won't be able to submit the form because the validation will never complete. 

I think that's what you're alluding to? I tried it this way first so know where that road leads haha 😄